{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"https://github.com/CropEvol/FAO_practice/blob/master/textbook/GenomicPrediction.ipynb","timestamp":1711957676692}],"toc_visible":true},"kernelspec":{"name":"python3","display_name":"Python 3"},"gpuClass":"standard"},"cells":[{"cell_type":"markdown","metadata":{"id":"vFnKcfhjYcjf"},"source":["<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/title.png?raw=true\" alt=\"title\" height=\"250px\">"]},{"cell_type":"markdown","metadata":{"id":"31NfnQcvPDoY"},"source":["# Genomic Prediction - practice session -\n","\n","In this notebook, we will perform Genomic Prediction analysis using real dataset (but unpublished!).\n","\n","And we will experience what can we do with genomic prediction models for breeding."]},{"cell_type":"markdown","source":["## The contents in this notebook ...\n","\n","* Make Genomic prediction model using real dataset\n","\n","  * We use rice population & grain number phenotype.\n","\n","* Application of genomic prediction model\n","  * We will consider good genotype for grain number by prediction model\n","  * We will search good parental lines to generate lines with high phenotype"],"metadata":{"id":"wZpUAKrgmHnm"}},{"cell_type":"markdown","source":["# Main contents"],"metadata":{"id":"pYgp_cqzl-pM"}},{"cell_type":"markdown","source":["## Mut-Map, QTL-seq, GWAS analysis\n","\n","We studied the methods to identify genetic variants associated with phenotypes such as Mut-map, QTL-seq.\n","\n","These approaches are focusing on **the identification of QTLs/genetic variants**.\n","\n","So, these approaches are very useful for **marker assisted selection** in breeding.\n","\n","Several crops were improved by introducing QTLs related yields, immunity, etc.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_1.png?raw=true\" alt=\"title\" height=\"150px\">\n","\n","However, most agronomical traits might be controlled by multiple genes, not only one gene.\n","\n","In addition, it is difficult to estimate effect of each gene by the identification methods like Mut-map, QTL-seq, and GWAS ...etc.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_2.png?raw=true\" alt=\"title\" height=\"300px\">\n","\n","If it is possible to estimate the effects of genetic variants, we can introduce genetic variants to breeding strategy more efficiently.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_3.png?raw=true\" alt=\"title\" height=\"140px\">\n","\n","**Genomic Prediction** is one of the approaches to estimate genetic effect of variants on phenotypes and predict phenotype values from genotype information.\n","\n","In this session, we'll introduce **what is Genomic Prediction model** and **what can be able to do by Genomic Prediction model**.\n","\n"],"metadata":{"id":"aifrycn-qcez"}},{"cell_type":"markdown","source":["# Review of Genomic Prediction\n","\n","Genomic prediction model is the approach to try to predict phenotype value by genotype information.\n","\n","Usually, we use SNP genotypes as input genotype data.\n","\n","So, in other words, genomic prediction model predicts phenotype value based on SNP genotypes.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/genomic_prediction.png?raw=true\" alt=\"colab\" height=\"300px\">\n","\n","To generate genomic prediction model, we have to understand the relationship between genotype and phenotype.\n","\n","So, we need the information about various genotypes and their phenotypes.\n","\n","We have to use a large segregating population and genotype & phenotype data of them to catch this relationship.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_19.png?raw=true\" alt=\"title\" height=\"250px\">\n","\n","\n","Then, we apply statistical methods to make the model that explain the relationship between genotype and phenotype.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_4.png?raw=true\" alt=\"title\" height=\"170px\">\n","\n","That's the brief process to make genomic prediction model.\n","\n","Let's experience the process to make genomic prediction model Step by Step."],"metadata":{"id":"F4tPUz-Y9TgH"}},{"cell_type":"markdown","source":["# Experience Genomic Prediction\n","In this section, we try to make genomic prediction model of rice.\n","\n","We will make prediction model that can predict **grain number** phenotype from genotype.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_21.png?raw=true\" alt=\"title\" height=\"150px\">\n","\n"],"metadata":{"id":"Xt6EApFB9zM2"}},{"cell_type":"markdown","source":["## **Genetic diverse population**\n","\n","At first, we have to generate genetic diverse  population to catch the relationship between genotype and phenotype.\n","\n","To make it easier to understand the relationship between phenotypes and genotypes,\n","\n","we often use **designed breeding population** as genetic diverse population\n"," for genomic prediction.\n","\n","Designed breeding population is like **RIL population, NAM population, MAGIC population, ...etc**.\n","\n","\n","In this practice, we'll use NAM population of rice.\n","\n","NAM(Nested Association Mapping) population is consist from multiple Recombinant inbred lines(RILs).\n","\n","NAM population is derived from crossing diverse parents with a common reference parent.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_5.png?raw=true\" alt=\"title\" height=\"300px\">\n","\n","In this practice session, we will use NAM population of rice.\n","\n","You can download phenotype data and genotype data by below command.\n","\n"],"metadata":{"id":"tXk9NBkAcSp0"}},{"cell_type":"code","source":["%%bash\n","wget -q -O genotype.csv https://raw.githubusercontent.com/slt666666/FAO_lecture/main/FAO_2024/data/genotype.csv\n","wget -q -O phenotype.csv https://raw.githubusercontent.com/slt666666/FAO_lecture/main/FAO_2024/data/phenotype.csv\n","wget -q -O modules.py https://raw.githubusercontent.com/slt666666/FAO_lecture/main/FAO_2024/data/modules2.py"],"metadata":{"id":"jIniocaQhMrh"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["After finish downloads, you can see phenotype & genotype dataset.\n","\n","Phenotype data file is `phenotype.csv` and genotype data is `genotype data.csv`."],"metadata":{"id":"nQ1Wmzn4_Maj"}},{"cell_type":"markdown","source":["### Phenotype data\n","\n","Our NAM population contains **747 lines** and we use **Grain number** as phenotype in this practice.\n","\n","We will use mean value of 5 replicates as phenotype value.\n","\n","You can check some part of phenotype data using below code.\n","\n","(This is unpublished data, so we just named NAM_XXX for each line)"],"metadata":{"id":"Xn0bP0chhYKz"}},{"cell_type":"code","source":["import pandas as pd\n","phenotype = pd.read_csv(\"phenotype.csv\")\n","phenotype"],"metadata":{"id":"t9DpBjdqhmdd"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["### Genotype data\n","\n","To generate Genotype data, we have to perform sequencing, alignment, and identification of SNPs as  experienced with Mut-map or QTL-seq analysis.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_6.png?raw=true\" alt=\"title\" height=\"300px\">\n","\n","But we omitted this process in this practice and we'll use the genotype data that has been organized into table data by our side.\n","\n","SNP genotypes were converted into **Number**.\n","\n","- Homozygous mutated genotype = 2\n","- Heterozygous mutated genotype = 1\n","- Original (Not mutated) genotype = 0\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_7.png?raw=true\" alt=\"title\" height=\"250px\">\n","\n","You can check some part of genotype data using below code."],"metadata":{"id":"98HviEWVln1D"}},{"cell_type":"code","source":["import pandas as pd\n","genotype = pd.read_csv(\"genotype.csv\")\n","genotype"],"metadata":{"id":"iJnr7zlQln-z"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Now, we got genotype & phenotype dataset.\n","\n","Next step is making genomic prediction model that explain the relationship between genotype and phenotype data.\n","\n","## **Make Genomic prediction model**\n","\n","From genotype and phenotype datasets,  we make statistical model that explain the relationship between genotype and phenotype.\n","\n","ex) very simle example\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_8.png?raw=true\" alt=\"title\" height=\"300px\">\n","\n","Recently, statistical analysis and machine learning methods showed remarkable progress.\n","\n","A lot of tools are available to make various models.\n","\n","Therefore, we can build the model very flexibly depending on the purpose,\n","\n","such as a model that assumes gene-gene interactions or a model include environmental factors into account.\n","\n","<br>\n","\n","In this practice, we will use one of the linear regression models as a simple example.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_9.png?raw=true\" alt=\"title\" height=\"70px\">\n","\n","Linear regression model can be calculated by [scikit-learn library](https://scikit-learn.org/stable/) that contains a lot of data analysis tools.\n","\n","```\n","â€» In this practice, we will not explain the detail of statistical things, data analysis, how to write code.\n","But, if you are interested in it, many lectures are available via online like: https://pll.harvard.edu/subject/data-science\n","```"],"metadata":{"id":"HddeA7RAvlnI"}},{"cell_type":"code","source":["from modules import linear_model\n","model = linear_model(genotype.iloc[:, 2:].T, phenotype)"],"metadata":{"id":"D9p42XRXqB10"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["We succeeded to make genomic prediction model.\n","\n","So, now, we can predict grain number from SNP genotypes using this model.\n","\n","However, we can't judge this model is good or bad.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_10.png?raw=true\" alt=\"title\" height=\"120px\">\n","\n","So, we have to prepare the approch to check the accuracy of the model.\n","\n","To do so, when we generate genomic prediction model,\n","\n","we usually separate dataset into training data and test data before making prediction model.\n","\n","1. Based on training data, we generate prediction model.\n","\n","1. Then, we'll check the accuracy of generated model using the rest of dataset (test data).\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_11.png?raw=true\" alt=\"title\" height=\"320px\">\n","\n","In our case, we make genomic prediction model using 597 lines (training data).\n","\n","Then, we will check the accuracy of generated model using 150 lines (test data).\n","\n","Let's check \"how accurate the model can be generated?\"."],"metadata":{"id":"uQZcjW9VqCGi"}},{"cell_type":"markdown","source":["### **1. Split dataset**\n","\n","At first, we split dataset into training data & test data.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_14.png?raw=true\" alt=\"title\" height=\"300px\">\n","\n","[scikit-learn library](https://scikit-learn.org/stable/) has data split function that split dataset so that the distribution of phenotype is the same."],"metadata":{"id":"oaq6rNW2kicm"}},{"cell_type":"code","source":["from sklearn.model_selection import train_test_split\n","import matplotlib.pyplot as plt\n","X_train, X_test, y_train, y_test = train_test_split(genotype.iloc[:, 2:].T, phenotype, test_size=0.2, random_state=1024)\n","plt.figure(figsize=[4,4])\n","plt.hist(y_train.Grain_number, label=\"Training data\")\n","plt.hist(y_test.Grain_number, label=\"Test data\")\n","plt.xlabel(\"Grain number\"); plt.ylabel(\"Lines\"); plt.legend(); plt.show()"],"metadata":{"id":"C7C4En3WwByK"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["We will use training data for prediction model and test data to check accuracy.\n","\n","### **2. Make prediction model**\n","\n","Then, we will make genomic prediction model based on training dataset.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_15.png?raw=true\" alt=\"title\" height=\"300px\">\n","\n","\n","We will make prediction model that explain the relationship between genotype and phenotype of training dataset by linear regression model using [scikit-learn library](https://scikit-learn.org/stable/)\n"],"metadata":{"id":"UofuExLom1tM"}},{"cell_type":"code","source":["from modules import linear_model\n","model = linear_model(X_train, y_train)"],"metadata":{"id":"SFlroDmUnVzo"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["We obtained linear regression model that explained the relationship between genotype & phenotype dataset.\n","\n","### **3. Check accuracy of the model**\n","\n","To check accuracy, we apply generated model in 2. to test dataset.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_16.png?raw=true\" alt=\"title\" height=\"300px\">\n","\n","We predict phenotype values based on genotype data of test data using generated prediction model.\n","\n","And compare observed phenotype values and predicted phenotype values.\n","\n","If these values are very similar, the model can predict phenotype values accurately.\n","\n","We can predict phenotype values from genotype information using generated prediction model by the below code."],"metadata":{"id":"iMPZ2Cfln4hh"}},{"cell_type":"code","source":["from modules import check_accuracy\n","y_test[\"Predicted_Grain_number\"] = model.predict(X_test)\n","display(y_test)\n","check_accuracy(model, X_test, y_test)"],"metadata":{"id":"o5MHlRmzpzqu"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The below scatterplot shows the relationship between observed and predicted phenotype values.\n","\n","Samples with low phenotypic values also tend to have low predicted phenotypic values.\n","\n","Samples with high phenotypic values also tend to have high predicted phenotypic values.\n","\n","The correlation coefficient between predicted phenotypes and observed phenotypes is most commonly used as indicator of prediction accuracy.\n","\n","In this case, correlation coefficient is 0.85~. It's not perfect prediction, but it seems like we can catch trends.\n","\n","Therefore, we succeeded to generate good genomic prediciton model and now we can predict grain number phenotype based on genotypes with high accuracy.\n","\n","<br>\n","\n","I omitted many process, but this is the general flow to make genomic prediction model and check accuracy of the model.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_11.png?raw=true\" alt=\"title\" height=\"250px\">\n"],"metadata":{"id":"HNajK2cDwL1L"}},{"cell_type":"markdown","source":["# Application of genomic prediction model\n","\n","Since we succeeded to generate genomic prediction model, we'll consider how to apply this model in breeding."],"metadata":{"id":"WAI2X2qTBlre"}},{"cell_type":"markdown","source":["## **Consider good genotype for traits**\n","\n","Genomic prediction model can predict phenotype values based only on genotype information.\n","\n","So, we can consider what genotype is ideal for traits.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_12.png?raw=true\" alt=\"title\" height=\"120px\">\n","\n","\n","In this session, let's try to make good genotype that improve grain number using genomic prediction model."],"metadata":{"id":"Z6yyI5M3CAQ3"}},{"cell_type":"markdown","source":["### Predict phenotype of customized genotype\n","\n","We can predict phenotype values of hypothetical genotype.\n","\n","For example, if we introduced another cultivar genotype in chromosome 1 & chromosome 3, how does phenotype change?\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_13.png?raw=true\" alt=\"title\" height=\"360px\">\n","\n","We can predict the phenotype value of this genotype by genomic prediction model.\n","\n","The below code predict phenotype values of hypothetical genotypes.\n","\n","Please check prediction results by below code."],"metadata":{"id":"efOCS7T2MG8F"}},{"cell_type":"code","source":["from modules import predict_customized_genotype\n","regions = [['chr01', 10000000, 20000000], ['chr03', 10000000, 20000000]]\n","predict_customized_genotype(genotype, regions, model)"],"metadata":{"id":"8TVxL8rMMLoJ"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Now, we can predict the phenotype values of new genotype.\n","\n","So, we can investigate the what kind of genotype generate high phenotype values.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_20.png?raw=true\" alt=\"title\" height=\"360px\">\n","\n","### **Play with genomic prediction**\n","\n","Let's try to find out genotypes with high grain number phenotype!\n","\n","The below code predict phenotype values that your customized genotype.\n","\n","You can edit the second line to specify genomic regions that you wan to introduce mutations (other cultivar's genotypes) as above example showed.\n","\n","`regions = [['chr01', 0, 20000000], ['chr03', 10000000, 20000000]]`.\n","\n","You can specify more than 2 regions like this:\n","\n","`regions = [ ['chrXX', YY, ZZZ], ['chrXX', YY, ZZZ], ['chrXX', YY, ZZZ], ['chrXX', YY, ZZZ] ]`\n","\n","For example, if you edit it to:\n","\n"," `regions = [['chr01', 12000000, 20000000], ['chr05', 0, 5000000], ['chr12', 10000000, 15000000]]`\n","\n","It means, another cultivar genotype is introduced into these â†“ genomic regions\n","* chr01...12 Mbp ~ 20 Mbp\n","* chr05...0 ~ 5 Mbp\n","* chr12...10 Mbp ~ 15 Mbp\n","\n","```\n","Common mistakes:\n","â€» Forget  ' ' of 'chr01' and  Forget 0 of 'chr01'\n","â€» Forget comma\n","â€» The number of [ ]\n","```\n","Let's try to find the genotype that produces a high grain number phenotypes while being careful of the common mistakes mentioned above!\n","\n","\n","â€»chromosome length is around 20000000bp ~ 40000000bp.\n"],"metadata":{"id":"Zp7LA-q_crpf"}},{"cell_type":"code","source":["from modules import predict_customized_genotype\n","regions = [['chr01', 10000000, 20000000], ['chr03', 10000000, 20000000]]\n","predict_customized_genotype(genotype, regions, model)"],"metadata":{"id":"XWxT4IK0c9Zz"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Based on this information, we can consider the direction of new cultivars to be created through breeding."],"metadata":{"id":"UBqYcrUFhz38"}},{"cell_type":"markdown","source":["## Genomic Selection: Consider good parental lines to generate progenies with high phenotypes\n","\n","If we have genotype inofrmation of breeding lines, we can simulate genotypes of next generation and predict phenotype values of them.\n","\n","So, we possibly identify the best parental pair to generate progenies with high phenotypes that would have been missed by phenotypic selection.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_17.png?raw=true\" alt=\"title\" height=\"400px\">\n","\n","<br>\n","\n","In our case, we have 747 lines of NAM population.\n","\n","<br>\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_5.png?raw=true\" alt=\"title\" height=\"300px\">\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/GP_18.png?raw=true\" alt=\"title\" height=\"240px\">\n","\n","To select good combination of parental lines from this population, genomic prediction model is useful.\n","\n","We can simulate genotypes of F2 population based on paretnal lines and predict phenotype values of F2 population using below code.\n","\n","\n","`You can change \"NAM_001\" & \"NAM_002\" to your favorite NAM lines (NAM_001 ~ NAM_747).`"],"metadata":{"id":"6q2CgMaP03QX"}},{"cell_type":"code","source":["from modules import predict_progeny_phenotype\n","predict_progeny_phenotype(\"NAM_001\", \"NAM_002\", 100, phenotype, genotype, model)"],"metadata":{"id":"z5tiu9uYgrUF"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["From this result, we can select the best combination of parental lines to make new cultivars."],"metadata":{"id":"ZDk-AaFfgrv8"}},{"cell_type":"markdown","metadata":{"id":"PFTQIMDGuYAT"},"source":["---\n","## Summary\n","\n","In this notebook, we play with **Genomic Prediction** analysis using unpublished data.\n","\n","We can predict phenotypes from genotype information by genomic prediction model.\n","\n","Thus, you can find out the ideal genotype for traits and good combination of parental lines.\n","   \n","If you have plan to generate **designed breeding population** which has high genetic variaty,\n","\n","genomic prediction approach is one of the approaches to achieve generating high-yield cultivar.\n","\n"]},{"cell_type":"code","source":[],"metadata":{"id":"qZnCziD1np5v"},"execution_count":null,"outputs":[]}]}