{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"https://github.com/CropEvol/FAO_practice/blob/master/textbook/QTL_seq.ipynb","timestamp":1711618171627}],"toc_visible":true},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"vFnKcfhjYcjf"},"source":["<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/title.png?raw=true\" alt=\"title\" height=\"250px\">\n"]},{"cell_type":"markdown","metadata":{"id":"31NfnQcvPDoY"},"source":["# QTL-seq - practice session -\n","\n","In this session, we will perform **QTL-seq analysis** using simulation data.\n","\n","In addition, we also perform **sliding window analysis** using real data (published data).\n","\n","Through this session, it may help you to understand ...\n","* The process of QTL-seq analysis\n","* How to interpret the results of QTL-seq analysis\n","* What kind of data is required for QTL-seq\n","* What is sliding window analysis & Why it is required."]},{"cell_type":"markdown","source":["## The contents in this notebook ...\n","\n","* QTL-seq analysis\n","  * Limitations of Mut-map analysis\n","  * QTL-seq analysis using simulation data (Step by Step)\n","  * Review of QTL-seq\n","\n","* Sliding window analysis\n","  * Limitations of the interpretation of the results of MutMap & QTL-seq analysis.\n","  * Sliding window analysis using published data."],"metadata":{"id":"wZpUAKrgmHnm"}},{"cell_type":"markdown","source":["# Main contents"],"metadata":{"id":"pYgp_cqzl-pM"}},{"cell_type":"markdown","source":["## **Limitations of Mut-map analysis**\n","\n","In the previous practice session, we perform Mut-map analysis to identify genetic variant that is associated with leaf color.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/Mutmap_8.png?raw=true\" alt=\"title\" height=\"200px\">\n","\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/Mutmap_12.png?raw=true\" alt=\"title\" height=\"250px\"><br>\n","\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/Mutmap_13.png?raw=true\" alt=\"title\" height=\"150px\">\n","\n","It looks this approach can be applied to wide variety traits.\n","\n","However, Mut-map analysis can be applied to only **Qualitative traits**.\n","\n","If traits were **Quantitative traits**, Mut-map can't be applied.\n","\n","Because, in case of Quantitative traits, segregating population (ex. F2 population) often showed a wide spectrum of variance.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/QTLseq_1.png?raw=true\" alt=\"title\" height=\"500px\">\n","\n","\n","And a lot of important agronomical traits are quntitative.\n"," - grain number\n"," - panicle number\n"," - root length\n"," - ...etc\n","\n","Also, Quantitative trait is controlled by not only few QTL, but also several QTLs.\n","\n"," To cope with these problem, we developped QTL-seq analysis.\n","\n"," QTL-seq is very similar method to Mut-map analysis, but QTL-seq is focusing on quantitative traits.\n","\n","\n"," **So, let's try to perform QTL-seq analysis step by step!**"],"metadata":{"id":"hbRDStM--gcz"}},{"cell_type":"markdown","source":["### **Practice of QTL-seq analysis**\n","\n","In this section, we will practice QTL-seq analysis under the following situation.\n","\n","- Reference genome of Cultivar B is available.\n","- Mutated cutivar showed high plant height compared to original cultivar.\n","- We generated 100 F2 progeny by crossing Cultivar B and mutated cultivar.\n","\n","**Purpose: Identify genetic variants that are associated with plant height using QTL-seq.**\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/QTLseq_2.png?raw=true\" alt=\"title\" height=\"500px\">\n","\n","Plant height is a quantitative trait, so F2 population showed wide variety of plant height.\n","\n","And we **can't apply Mut-map analysis.**\n","\n","Let's check the distribution of phenotype values of this F2 population.\n","\n","Below code download the phenotype data of F2 population and visualize it."],"metadata":{"id":"-bmR8RtjIghn"}},{"cell_type":"code","source":["import pandas as pd\n","import matplotlib.pyplot as plt\n","import seaborn as sns\n","!wget -O F2_phenotype.csv https://raw.githubusercontent.com/slt666666/FAO_lecture/main/FAO_2024/data/F2_phenotype.csv 2>/dev/null\n","phenotype = pd.read_csv(\"F2_phenotype.csv\")\n","fig, ax = plt.subplots(figsize=(4, 3))\n","sns.histplot(phenotype[\"plant height\"], bins=20, ax=ax)\n","plt.show()"],"metadata":{"id":"2z1-S0KkKE_p"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["In our F2 population showed the wide distribution of plant height.\n","\n","Basically, quantitative traits are controlled by multiple genes.\n","\n","Therefore, phenotype values are variable depending on the combination of genetic variants.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/QTLseq_3.png?raw=true\" alt=\"title\" height=\"500px\">\n","\n","In this case, F2 progenies with taller plant height are likely contain genetic variants with strong positive effect (ex. the 2nd SNP in above figure).\n","\n","In other words, F2 progenies with shorter plant height are likely do **not** contain genetic variants with strong positive effect.\n","\n","So, genetic varinats with strong positive effect are conserved to F2 progenies with tall plant height.\n","\n","QTL-seq analysis use this biased distribution of genetic variants. (The idea is similar to MutMap analysis.)\n","\n","Therefore, QTL-seq analysis is the approach to identify genetic variants that **conserved in F2 progenies with high plant height** and **did not conserved F2 progenies with low plant height** to identify causal SNPs."],"metadata":{"id":"jmTNGFoxJ-R1"}},{"cell_type":"markdown","source":["#### **Bulked DNA sequencing**\n","\n","In QTL-seq analysis, we want to identify SNPs that are mutated in the most F2 progenies with taller plant height and not mutated in the most F2 progenies with shorter plant height.\n","\n","So, in the QTL-seq anaysis, we perform 2 bulked sequencing.\n","\n","First, we extract DNA from F2 progenies with top 10 plant height and bulked sequencing.\n","\n","We also extract DNA from F2 progenies with bottom 10 plant height and bulked sequencing.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/QTLseq_4.png?raw=true\" alt=\"title\" height=\"450px\">\n","\n","Sequencing data and Bioinformatics tools can be downloaded using below command."],"metadata":{"id":"uHcn0bkd7Rob"}},{"cell_type":"code","source":["%%bash\n","# download sequence data\n","wget https://raw.githubusercontent.com/slt666666/FAO_lecture/main/FAO_2024/data/get_data2.sh 2>/dev/null\n","bash get_data2.sh\n","\n","# install bwa, samtools, bcftools\n","apt-get install -q bwa\n","apt-get install -q samtools\n","apt-get install -q bcftools\n","\n","## igv-notebook-0.3.1\n","pip install -q igv-notebook==0.3.1\n","wget -q -O modules.py https://raw.githubusercontent.com/slt666666/FAO_lecture/main/FAO_2024/data/modules.py\n","\n","# In this practice, the installation command is tailored to Google Colab.\n","# When you try your own PC, please check the installation method on each tool's website."],"metadata":{"id":"R2OZ2AGsgj7t"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Bulked sequencing data is `reads/high_bulked_read1/2.fastq` and `reads/low_bulked_read1/2.fastq`.\n","\n","These sequencing data is derived from multiple F2 progenies with high plant height and low plant height.\n","\n","So, if we perform alignment using these sequences, we can identify causal muation that is conserved F2 progenies with high plant height.\n","\n","#### **Alignment**\n","\n","Next step is the alignment by `bwa`.\n","\n","To identify genetic variants that are conserved in high-phenotype F2 progenies and not conserved in low-phenotype F2 progenies,\n","\n","we will compare alignment results of high & low bulked sequencing.\n","\n","Let's see what kind of alignment results we get when we use **2** bulked sequences."],"metadata":{"id":"5RQlwtYe7BZW"}},{"cell_type":"code","source":["%%bash\n","# alignment high bulked sequencing data to referece genome\n","bwa index genome/CultivarB.fa\n","bwa aln genome/CultivarB.fa reads/high_bulked_read1.fastq > high_bulked_read1.sai 2>/dev/null\n","bwa aln genome/CultivarB.fa reads/high_bulked_read2.fastq > high_bulked_read2.sai 2>/dev/null\n","bwa sampe genome/CultivarB.fa high_bulked_read1.sai high_bulked_read2.sai reads/high_bulked_read1.fastq reads/high_bulked_read2.fastq > high_bulked.sam; rm -fR high_bulked_read1.sai high_bulked_read2.sai\n","# convert sam to bam\n","samtools sort -O bam high_bulked.sam > high_bulked.bam\n","samtools index high_bulked.bam\n","\n","# alignment low bulked sequencing data to referece genome\n","bwa index genome/CultivarB.fa\n","bwa aln genome/CultivarB.fa reads/low_bulked_read1.fastq > low_bulked_read1.sai 2>/dev/null\n","bwa aln genome/CultivarB.fa reads/low_bulked_read2.fastq > low_bulked_read2.sai 2>/dev/null\n","bwa sampe genome/CultivarB.fa low_bulked_read1.sai low_bulked_read2.sai reads/low_bulked_read1.fastq reads/low_bulked_read2.fastq > low_bulked.sam; rm -fR low_bulked_read1.sai low_bulked_read2.sai\n","# convert sam to bam\n","samtools sort -O bam low_bulked.sam > low_bulked.bam\n","samtools index low_bulked.bam\n","\n","# generate vcf file\n","bcftools mpileup -Ob -o bulked.bcf -f genome/CultivarB.fa --annotate FORMAT/AD low_bulked.bam high_bulked.bam\n","bcftools call -vmO v -o bulked.vcf bulked.bcf; rm bulked.bcf"],"metadata":{"id":"qaHxlTye_MYp"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["After finish the alignment, you will obtain `high_bulked.bam`, `low_bulked.bam`, `bulked.vcf` as a aignment results.\n","\n","Then, let's visualize alignment results for high bulked sequencing & low bulked sequencing."],"metadata":{"id":"ovepJGcl7v8R"}},{"cell_type":"code","source":["## Visualize alignment results\n","import igv_notebook\n","from modules import RefTrack, AnnotationTrack, BamTrack\n","igv_notebook.init()\n","# show reference genome\n","ref = RefTrack({ \"fastaPath\":\"genome/CultivarB.fa\", \"indexPath\":\"genome/CultivarB.fa.fai\", \"id\":\"CultivarB\" })\n","# show aligned high bulked reads\n","B = BamTrack({ \"name\":\"high_bulked\", \"path\":\"high_bulked.bam\", \"indexPath\":\"high_bulked.bam.bai\", \"viewAsPairs\":True })\n","# show aligned low bulked reads\n","B2 = BamTrack({ \"name\":\"low_bulked\", \"path\":\"low_bulked.bam\", \"indexPath\":\"low_bulked.bam.bai\", \"viewAsPairs\":True })\n","# visualize\n","b = igv_notebook.Browser(ref)\n","b.load_track(B)\n","b.load_track(B2)"],"metadata":{"id":"1ZUEum8P_fPT"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["We can check genetic variants conserved in high(low) phenotype F2 progenies.\n","\n","#### **SNP index**\n","\n","As a results, we obtained SNPs from each bulked sequences.\n","\n","From these SNPs, we want to identify SNPs that is conserved high phenotype F2 progenies.\n","\n","This kind of SNPs can be identified based on SNP index that was also used in Mut-map analysis.\n","\n","For example ↓\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/QTLseq_5.png?raw=true\" alt=\"title\" height=\"300px\">\n","\n","All high bulk samples have mutated genotype of this SNP.\n","\n","But, we can't say this SNP is causal genetic variant yet.\n","\n","We have to consider **low bulk samples**.\n","\n","Even if the SNP index is very high in high bulk, it can't be considered a causative genetic variants if it is **also high SNP index in low bulk**.\n","\n","(like left SNP in figure).\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/QTLseq_6.png?raw=true\" alt=\"title\" height=\"450px\">\n","\n","Therefore, we have to identify SNPs with high SNP index in high bulk, but low SNP index in low bulk.\n","\n","(like right SNP in figure).\n","\n","#### **ΔSNP index**\n","\n","To identify this kind of SNPs, we'll use **ΔSNP index**.\n","\n","\n","**ΔSNP index** is calculated by `SNP index(high bulk) - SNP index(low bulk)`.\n","\n","**High ΔSNP index means only high bulk F2 progenies have mutated genotype, and few or no low bulk F2 progenies have it.**\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/QTLseq_7.png?raw=true\" alt=\"title\" height=\"450px\">\n","\n","In summary, we have to identify **SNPs with high ΔSNP index.** \\\n","\n","This SNP is only conserved in high phenotype F2 progenies, so it could be causative variants.\n","\n","Let's calculate ΔSNP index from vcf file (alignment results).\n","\n","You can calculate ΔSNP index values by my script using below code.\n"],"metadata":{"id":"Sv2k0jHYbEfb"}},{"cell_type":"code","source":["from modules import calc_delta_SNP_index\n","delta_SNP_index = calc_delta_SNP_index(\"bulked.vcf\")\n","delta_SNP_index"],"metadata":{"id":"BJoYGBTqlMD5"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["It's not so difficult to manually check ΔSNP index if the number of SNPs is few hundreds.\n","\n","But if you find more SNPs like thousands to millions...etc, it is better to visualize it with a graph.\n","\n","Let's try to visualize SNP index & ΔSNP index using below code."],"metadata":{"id":"FUsq4E6bsmij"}},{"cell_type":"code","source":["from modules import visualize_delta_SNP_index\n","visualize_delta_SNP_index(delta_SNP_index)"],"metadata":{"id":"YajQ_W_QmtA9"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["You can see SNP index of high/low bulk and ΔSNP index across whole genome.\n","\n","Basically, genetic variants randomly mixed by recombinations in F2 population.\n","\n","So, SNP index of high bulk and low bulk is around 0.5, and ΔSNP index of a **not** causal SNP is aroud 0 (red line).\n","\n","On the other hand, only high bulk F2 progenies have causal SNP genotype.\n","\n","So, ΔSNP index of a causal SNP is close to 1. In this case, **a SNP located around 5,800 bp might be a causal genetic variant**.\n","\n","Also, we can identify several SNPs showed ΔSNP index > 0. **These SNPs might also have positive effect on phenotypes.**\n","\n","(because not all, but most high phenotype samples have them).\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/QTLseq_8.png?raw=true\" alt=\"title\" height=\"450px\">\n","\n","<br>\n","\n","**This is the process of QTL-seq analysis!**"],"metadata":{"id":"CsyWj6lMrWwI"}},{"cell_type":"markdown","source":["# Review of QTL-seq\n","\n","Let's take a look at whole process of QTL-seq analysis.\n","\n","  QTL-seq analysis is one the methods to identify the genomic region which is associate with the quantitative traits. (like QTL mapping.)\n","\n","* MutMap: Qualitative trait\n","\n","* QTL-seq: Quantitative trait\n","\n","To perform QTL-seq analysis...\n","1. Generate a segregating population like F2 population by crossing the mutated cultivar and the original cultivar.\n","  - By creating an F2 population, we obtain a large number of individuals with shuffled genome.\n","  - Each mutation introduced to the genome has about a 50% chance of being transmitted to each F2 individual.\n","\n","1. If there are segregtion as a quantitative trait, select 10~ samples each showing opposite phenotypes (high and low).\n","\n","  - The F2 individuals with the high phenotype should have common causal genetic variants in the genome.\n","  - And individuals with the low phenotype should **not** have this causal variant.\n","\n","1. Collect bulked DNA from selected samples for each phenotype (Low bulk & High bulk), and perform sequencing (bulk sequencing).\n","\n","1. Alignment of bulked sequences to the reference genome. You can get information like\n","  - The position where the mutation\n","  - The number of reference and mutant bases in the high/low bulked sequence, respectively\n","\n","1. Finally, to compare SNP-index between high bulk and low bulk, calculate ΔSNP-index which indicate the difference of SNP-index between high and low bulks.\n","  - SNPs which allele frequency is hugely different between high and low bulk are identified as high ΔSNP-index(0 <) or low ΔSNP-index(< 0).\n","\n","This genetic vatinat is candidate of causative variant that affect phenotype.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/QTLseq_9.png?raw=true\" alt=\"title\" height=\"750px\">\n"],"metadata":{"id":"ISbXd9Yh93Tv"}},{"cell_type":"markdown","source":["# Introduction of QTL-seq pipeline\n","\n","In this practice session, we perform QTL-seq analysis step by step using bioinformatics tools like bwa, samtools, bcftools...etc and our own scripts.\n","\n","Our laboratory developped the very simple pipeline to conduct QTL-seq.\n","\n","<img src=\"https://github.com/YuSugihara/QTL-seq/blob/master/images/1_logo.png?raw=true\" alt=\"title\" height=\"100px\">\n","\n","(https://github.com/YuSugihara/QTL-seq)\n","\n","This pipeline is very simple to use. Required input data is just sequencing data (= fasta and fastq files).\n","\n","And you just install pipeline & use below command. That's it !\n","\n","```\n","# command\n","qtlseq -r reference.fasta -p reference.fastq -b1 high_bulked_sequences.fastq -b2 low_bulked_sequences.fastq -n1 20 -n2 20 -o output_name\n","```\n","\n","Then, you can get data files that contains SNP positions & SNP-index values & graph that visualize it.\n","\n","So, if you are interested in perform QTL-seq analysis, this pipeline is one option."],"metadata":{"id":"DT0Rb-vsyrZH"}},{"cell_type":"markdown","source":["# **Sliding window analysis**\n","\n","Finally, we introduce **Sliding window analysis** for Mut-map & QTL-seq.\n","\n","In this part, we use yesterday's MutMap result of published data to experience sliding window analysis.\n"],"metadata":{"id":"ZbisCkqHNt8C"}},{"cell_type":"markdown","source":["## **Review of Sliding Window analysis**\n","\n","Different from simulation study, the real data (experimental data) does not always showed theoretically correct values.\n","\n","The real data usually contains **sequencing errors, outliers, human error...etc**.\n","\n","And it generates **false positives.**\n","\n","So, in case of Mut-Map and QTL-seq analysis **using real dataset**, SNP index will be different from expected values due to the errors.\n","\n","Especially, when the amount of sequencing data or population size is not enough, it usually happens.\n","\n","Sometimes it could lead a big outlier like SNP index=1 in the not important SNP position.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/Slidingwindow_1.png?raw=true\" alt=\"title\" height=\"250px\">\n","\n","### **Sliding window**\n","\n","To escape these errors, one approach is taking the average of SNP indexes of nearby regions.\n","\n","Considering the linkage, the SNP index of nearby regions should have similar values.\n","\n","So, sliding window will diminish the influence of error, and we can get more reliable SNP index results.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/Slidingwindow_4.png?raw=true\" alt=\"title\" height=\"300px\">\n","\n","<br>\n","\n","The process of sliding window analysis is...\n","\n","1. Extract data (SNP-indexes) in the interval regions (of window size).\n","1. Calculate average value of them.\n","1. Move the interval by the step size.\n","1. Continue 2-4 process to finish the whole region.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/Slidingwindow_2.png?raw=true\" alt=\"title\" height=\"400px\">\n","\n","Let's see how much it differs from the original SNP index, **after sliding window analysis.**"],"metadata":{"id":"95A7GRBEsOjS"}},{"cell_type":"markdown","source":["## Example\n","\n","In this section, we try to perform sliding window analysis using published data.\n","\n","We used published results of SNP index of Mut-map analysis as an example [(Abe et al., 2012)](https://www.nature.com/articles/nbt.2095).\n","\n","This paper try to identify genetic variants that are associated with leaf color (green to lightgreen).\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/material.png?raw=true\" alt=\"title\" height=\"400px\">\n","\n","### SNP index\n","\n","At first, we try to visualize original SNP index values using below code."],"metadata":{"id":"tWZ-iVMD0foG"}},{"cell_type":"code","source":["import pandas as pd\n","from modules import visualize_SNP_index\n","\n","!wget -O published_Mutmap.csv https://raw.githubusercontent.com/slt666666/FAO_lecture/main/FAO_2024/data/published_Mutmap.csv 2>/dev/null\n","SNP_index = pd.read_csv(\"published_Mutmap.csv\")\n","visualize_SNP_index(SNP_index)"],"metadata":{"id":"p7yOprThSkOA"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["There are multiple SNPs showed high SNP index.\n","\n","From this plot, it's difficult to distinguish true positives and false positives.\n","\n","So, to extract more reliable regions from this results, we try to perform sliding window analysis for this result."],"metadata":{"id":"Si_y1YaxwmTt"}},{"cell_type":"markdown","source":["## Sliding window analysis & visualize it.\n","\n","We try to perform sliding window analysis by our script.\n","\n","```\n","Window size & Step size to perform sliding window analysis can be set.\n","At first, we set window size = 1mbp & step size = 200kbp\n","(considering that the legth of chromosome10 is 23207287bp.)\n","```"],"metadata":{"id":"PiCsEGlBw8Z1"}},{"cell_type":"code","source":["from modules import sliding_window\n","average_SNP_index = sliding_window(SNP_index, window_size=1000000, step_size = 200000)"],"metadata":{"id":"aGu4wSAQ5Kzh"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Red line showed average SNP-index for each interval.\n","\n","Plot showed right end position(22~23Mbp) in chromosome 10 showed high average SNP index.\n","\n","This result suggests that there is a gene that related to leaf color in 22~23Mbp, chromosome 10.\n","\n","In this region, there is a gene \"*OsCAO1*\" that codes Chlorophyllide a oxygenase.\n","\n","And it was known that the knock-out line of *OsCAO1* showed light green leaf. (Abe et al., 2012)。\n","\n","<br>\n","\n","But, of course, even after applying sliding window anaylsis, results may still contains false positive.\n","\n","In that case, we have to consider other approach to select candidate SNPs, such as biological information of the SNP position(exon, important gene region...etc) or increase population size / sequence data.\n"],"metadata":{"id":"9wguB-6sx0YF"}},{"cell_type":"markdown","source":["## Our developped MutMap/QTL-seq analysis pipeline include sliding window analysis\n","\n","In this practice, we perform sliding window analysis by our scripts.\n","\n","But the pipeline that we introduced before contains sliding window analysis.\n","\n","You can specify window size & step size in these pipelines.\n","\n","- MutMap\n","https://github.com/YuSugihara/MutMap#Usage\n","\n","- QTL-seq\n","https://github.com/YuSugihara/QTL-seq#usage"],"metadata":{"id":"2cTX5tWy2VlH"}},{"cell_type":"markdown","source":["## **ex) The result of QTL-seq analysis of published data**\n","\n","In case of QTL-seq analysis, sliding window is important.\n","\n","For example, [Takagi et al., 2012](https://doi.org/10.1111/tpj.12105) showed the identification of genetic variants associated with disease resistance.\n","\n","<img src=\"https://github.com/slt666666/FAO_lecture/blob/main/FAO_2024/Slidingwindow_3.png?raw=true\" alt=\"title\" height=\"650px\">\n","\n","Original ΔSNP index plot showed that multiple SNPs have high ΔSNP index (blue dot).\n","\n","But, we succeeded to identify causal genomic regions after sliding window analysis (red line).\n"],"metadata":{"id":"irIdvDCxvQ_Q"}},{"cell_type":"markdown","metadata":{"id":"PFTQIMDGuYAT"},"source":["---\n","## Summary\n","\n","In this notebook, we demonstrate **QTL-seq** analysis using simulation data to understand the process of QTL-seq analysis.\n","\n","Also, we introduced **Sliding window** analysis that is requred to identify reliable causal genetic wariants.\n","\n","<br>\n","\n","The important point is how to interpret the results (SNP-index values).\n","\n","We should take care of false positives, errors, ...etc by applying sliding window analysis, checking read depth of interested SNPs...etc\n","\n","<br>\n","\n","And if you want to conduct QTL-seq analysis, the pipeline that our group developped is prepared. (https://github.com/YuSugihara/QTL-seq)\n","\n","Next session, we'll intorduce **Genomic Prediction**."]}]}